File: src/org/kobjects/nlp/latin/whitaker/Converter.java
package org.kobjects.nlp.latin.whitaker;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Writer;

/**
 * Tool to convert the plain text dictionary output from whitaker's words to a simplified
 * format.
 */
public class Converter {

	public static String resolveCodes(String codes) {
		StringBuilder sb = new StringBuilder();
		switch (codes.charAt(0)) {
			case 'A': sb.append("age=archaic, "); break;
			case 'B': sb.append("age=early, "); break;
			case 'C': sb.append("age=classical, "); break;
			case 'D': sb.append("age=late, "); break;
			case 'E': sb.append("age=later, "); break;
			case 'F': sb.append("age=medieval, "); break;
			case 'G': sb.append("age=scholar, "); break;
			case 'H': sb.append("age=modern, "); break;
		}
		switch (codes.charAt(1)) {
			case 'A': sb.append("cat=agr, "); break;
			case 'B': sb.append("cat=bio, "); break;
			case 'D': sb.append("cat=drm, "); break;
			case 'E': sb.append("cat=ecc, "); break;
			case 'G': sb.append("cat=grm, "); break;
			case 'L': sb.append("cat=leg, "); break;
			case 'P': sb.append("cat=poe, "); break;
			case 'S': sb.append("cat=sci, "); break;
			case 'T': sb.append("cat=tec, "); break;
			case 'W': sb.append("cat=war, "); break;
			case 'Y': sb.append("cat=myt, "); break;
		}
		switch (codes.charAt(2)) {
		case 'A': sb.append("reg=Aftrica, "); break;
		case 'B': sb.append("reg=Britain, ");break;
		case 'C': sb.append("reg=China, ");break;
		case 'D': sb.append("reg=Scandinavia, ");break;
		case 'E': sb.append("reg=Egypt, ");break;
		case 'F': sb.append("reg=France/Gaul, ");break;
		case 'G': sb.append("reg=Germany, ");break;
		case 'H': sb.append("reg=Greece, ");break;
		case 'I': sb.append("reg=Italy/Rome, ");break;
		case 'J': sb.append("reg=India, ");break;
		case 'K': sb.append("reg=Balkans, ");break;
		case 'N': sb.append("reg=Netherlands, ");break;
		case 'P': sb.append("reg=Persia, ");break;
		case 'Q': sb.append("reg=NearEast, ");break;
		case 'R': sb.append("reg=Russia, ");break;
		case 'S': sb.append("reg=Spain/Iberia, ");break;
		case 'U': sb.append("reg=Eastern Europe, ");break;
		}
		
		switch (codes.charAt(3)) {
		case 'A': sb.append("frq=+++, ");break;
		case 'B': sb.append("frq=++, ");break;
		case 'C': sb.append("frq=+, ");break;
		case 'D': sb.append("frq=-, ");break;
		case 'E': sb.append("frq=--, ");break;
		case 'F': sb.append("frq=---, ");break;
		case 'I': sb.append("frq=inscription, ");break;
		case 'M': sb.append("frq=grafffiti, ");break;
		case 'N': sb.append("frq=Pliny, ");break;
		}
		switch (codes.charAt(4)) {
		case 'B': sb.append("src=Bee, ");break;
		case 'C': sb.append("src=CAS, ");break;
		case 'D': sb.append("src=Sex, ");break;
		case 'E': sb.append("src=Ecc, ");break;
		case 'F': sb.append("sec=Def, ");break;
		case 'G': sb.append("src=G+L, ");break;
		case 'H': sb.append("src=Ouvrard, ");break;
		case 'I': sb.append("src=Leverett1845, ");break;
		case 'K': sb.append("src=Cal, ");break;
		case 'L': sb.append("src=Levis1891, ");break;
		case 'M': sb.append("src=Latham1980, ");break;
		case 'N': sb.append("src=Nelson, ");break;
		case 'O': sb.append("src=OLD, ");break;
		case 'P': sb.append("src=Souter1949, ");break;
		case 'Q': sb.append("src=other, ");break;
		case 'S': sb.append("src=L+S, ");break;
		case 'T': sb.append("src=translation, ");break;
		case 'U': sb.append("src=DuCange, ");break;
		case 'V': sb.append("src=Saxo, ");break;
		case 'W': sb.append("src=guess, ");break;
		case 'Z': sb.append("src=user, ");break;
		}
		
		if (sb.length() > 0) {
			sb.setLength(sb.length() - 2);
		}

		return sb.toString();
	}
	
	public static String supertrim(String s) {
		StringBuilder sb = new StringBuilder(s.length());
		boolean wasSpace = true;
		for (int i = 0; i < s.length(); i++) {
			char c = s.charAt(i);
			if (c <= ' ') {
				if (!wasSpace) {
					sb.append(' ');
					wasSpace = true;
				} 
			} else {
				sb.append(c);
				wasSpace = false;
			}
		}
		if (wasSpace && sb.length() > 0) {
			sb.setLength(sb.length() - 1);
		}		
		return sb.toString();
	}
	
	public static String cleanKind(String kind) {
		kind = supertrim(kind);
		int cut0 = kind.indexOf('(');
		int cut1 = kind.indexOf(')', cut0 + 1);
		if (cut0 != -1 && cut1 > cut0 + 1) {
			char d = kind.charAt(cut0+1);
			if (d >= '0' && d <= '9') {
				kind = kind.substring(0,  cut0) + d + kind.substring(cut1 + 1);
			}
		}
		return kind;
	}
	
	
	public static void main(String args[]) throws Exception {
		BufferedReader reader = new BufferedReader(new InputStreamReader(Converter.class.getResourceAsStream("dictpage.txt"), "iso-8859-1"));
		
		Writer out = new OutputStreamWriter(new FileOutputStream(new File("src/org/kobjects/nlp/latin/whitaker_converted.txt")), "utf-8");
		
		String nextLine = reader.readLine();
		do {
			String line = nextLine;
			nextLine = reader.readLine();
			
			while (nextLine != null && nextLine.indexOf('|') != -1) {
				int cut = nextLine.lastIndexOf('|');
				line += " " + nextLine.substring(cut + 1);
				nextLine = reader.readLine();
			}
			
			if (!line.startsWith("#")) {
				continue;
			}
			
			int cut0 = line.indexOf("  ");
			int cut1 = line.indexOf("[", cut0);
			int cut2 = line.indexOf(" :: ");
			
			String words = line.substring(1, cut0).trim();
			String kind = line.substring(cut0 + 2, cut1).trim();
			String codes = line.substring(cut1 + 1, cut1 + 7);
			String translation = line.substring(cut2 + 4).trim();
			
			// See http://archives.nd.edu/whitaker/wordsdoc.htm#Dictionary%20Codes
			
			kind = cleanKind(kind);
			String decoded = resolveCodes(codes);
			
			out.write(words);
			out.write("; ");
			out.write(kind);
			if (decoded.length() > 0) {
				out.write("; ");
				out.write(decoded);
			}
			out.write(":\n");
			
			String[] parts = translation.split(";");
			for (int i = 0; i < parts.length; i++) {
				out.write("  ");
				out.write(parts[i].trim());
				out.write(i < parts.length - 1 ? ";\n" : "\n");
			}
			
		} while (nextLine != null);
		
		out.close();
		
	}
	
}

File: bin/Converter.js

/**
 * Tool to convert the plain text dictionary output from whitaker's words to a simplified
 * format.
 * 
 * CHANGES FROM ORIGINAL:
 * - Additional definitions (called "continuation MEAN" defs)
 *   from new lines starting with the "|" character are now included
 *
 * TODO:
 * - How are you supposed to support sum/esse etc?
 * - Process definitions marked with "[suffix => def]" if they aren't
 */
  
 import fs from "fs";
 import path from "path";
 import split from "split";
 import { Transform } from "stream";
 import CombinedStream from 'combined-stream';
 import { fileURLToPath } from "url";
 
 import { supertrim } from "../src/api/Strings.js";
 
 const __filename = fileURLToPath(import.meta.url);
 const __dirname = path.dirname(__filename);
 const inputDirPath = path.join(__dirname, "src");
 const outputFilePath = path.join(__dirname, "dest", "dictionary.txt");
 
 const readStream = CombinedStream.create();
 const readMainStream = fs.createReadStream(path.join(inputDirPath, "dictpage.txt"), "latin1");
 let readExtensionsStream = fs.createReadStream(path.join(inputDirPath, "extensions.lat"), "utf8");
 
 readStream.maxDataSize = 1024 * 1024 * 8; // 8 Megabytes
 
 // Stream event handlers
 const writeStream = fs.createWriteStream(outputFilePath, "utf-8");
 writeStream
   .on('close', () => {
     console.log('conversion finished.')
     process.exit();
   });
 readStream
   .on("error", (err) => { throw new Error(err) })
   .on('finish', () => {
     console.log(`Finished processing dictionary files in '${inputDirPath}'.\n Saved output to '${outputFilePath}'.`);
     readStream.destroy();
   })
   .on('close', () => writeStream.end());
 
 
let removalsArray = null;
function readRemovalsArray() {
  const removeFileContents = fs.readFileSync(path.join(inputDirPath, "remove.lat"), "latin1");
  return removeFileContents.substring(1).split("\n#");
}

const filterTransformStream = new Transform({
  transform(data, encoding, callback) {
    const line = data.toString();
    if (isLineExcluded(line)) {
      console.log("EXCLUDED", line);
      callback();
    } else {
      this.push(filterAndCutLine(data.toString()));
      callback();
    }
  },

  // Add extensions, none of which should be filtered
  flush(callback) {
    this.push(readExtensionsBuffer);
    console.log("Flushed")
    callback();
  }
})
const processTransformStream = new Transform({
  transform (data, encoding, callback) {
    console.log(data.toString())
    this.push(processLine(data.toString()));
    callback();
  }
});
 
//  readMainStream
//    .on("error", (err) => { throw new Error(err) })
//    .pipe(split())
//    .pipe(filterRemovalsStream)
//    .on("close", () => {
//      console.log("Finished")
//      readExtensionsStream
//        .pipe(split())
//        .on("finish", () => {
//          readStream.append(readMainStream)
//          readStream.append(readExtensionsStream);
//          convertAndWrite();
//        })
//    })
//    .on("close", () => console.log("CLOSED"))
 
// Prep other Streams
readStream.append(readMainStream);

// Todo: This is technically out of timing (needs to be ready on filterTransformStream's flush event),
// but it will finish before needed
const readExtensionsBufferChunks = [];
let readExtensionsBuffer = null;
readExtensionsStream.pipe(split())
  .on("data", chunk => readExtensionsBufferChunks.push(Buffer.from(chunk, "utf-8")))
  .on('end', () => {
    readExtensionsBuffer = Buffer.concat(readExtensionsBufferChunks)
  });

// Main Stream Process
readMainStream
  .on("error", (err) => { throw new Error(err) })
  .pipe(split())
  .pipe(filterTransformStream)
  .pipe(split(/(?:\:\:)|[\n;]/))
  .pipe(processTransformStream)
  .pipe(writeStream)
  .on("close", () => console.log("CLOSED"));

// readExtensionsStream
//   .on("error", (err) => { throw new Error(err) })
//   .pipe(split())
//   .on("finish", () => {
// })

// readStream.append(readMainStream)
// readStream.append(readExtensionsStream);
// convertAndWrite();

// Step 2: Convert and write to file
function convertAndWrite() {
  readStream
    .pipe(split())
    .pipe(filterTransformStream)
    .pipe(split(/(?:\:\:)|[\n;]/))
    .pipe(processTransformStream)
    .pipe(writeStream)
}

function isLineExcluded(line) {
  if (!removalsArray) {
    removalsArray = readRemovalsArray();
  }
  const words = line.substring(1, line.indexOf("  ")).split(", ")
  return words.some(word => (
    removalsArray.includes(word)
  ));
}
 
 function filterAndCutLine(line) {
   const pipeIndex = line.indexOf("|");
   
   // Case 1: Merge with prev line
   if (pipeIndex > -1) {
     return ` ${line.substring(pipeIndex + 1)}`;
   }
   
   // Case 2: skip line
   let cut0 = line.indexOf("  ");
   let cut1 = line.indexOf("[", cut0);        
   let kind = supertrim(line.substring(cut0 + 2, cut1));
   
   // Irregular words requiring extra kind data for case
   // Commented but left in in case something breaks by allowing these, unlike Java repo
   // Todo: Set extra data and pull that in Declinator/Conjugator
   //
   // if (
   //   kind.startsWith("V ") && (kind.indexOf("ACTIVE") !== -1 || kind.indexOf("PASSIVE") !== -1) ||
   //   (kind.startsWith("N")  && kind.length > 3) ||
   //   (kind.startsWith("ADJ") && kind.length > 3)
   // ) {
   //   return "";
   // }
   // Abbreviations
   if (kind.startsWith("abb.")) {
     return "";
   }
 
   // Case 3: Normal
   return line;
 }
 
 function processLine(line) {
   // Incorrect Regex 
   if (!line.trim()) {
     return "";
   // Dictionary definitions
   } else if (!line.startsWith("#")) {
     return ` ${line};\n`;
   }
   
   let cut0 = line.indexOf("  ");
   let cut1 = line.indexOf("[", cut0);
         
   const latin = supertrim(line.substring(1, cut0));
   let kind = supertrim(line.substring(cut0 + 2, cut1));
   let codes = supertrim(line.substring(cut1 + 1, cut1 + 7));
   // const english = supertrim(line.substring(cut2 + 4));
 
   kind = cleanKind(kind);
   codes = resolveCodes(codes);
 
   const codesPart = codes?.length ? `; ${codes}` : "";
 
   // This skips the semicolon for the last english definition - problem?
   //
   // String[] parts = english.split(";");
   // for (let i = 0; i < parts.length; i++) {
   //   out.write("  ");
   //   out.write(parts[i].trim());
   //   out.write(i < parts.length - 1 ? ";\n" : "\n");
   // }
 
   return `${latin}; ${cleanKind(kind)}${codesPart}:\n`;
 }
 
 
 // See http://archives.nd.edu/whitaker/wordsdoc.htm#Dictionary%20Codes
 function resolveCodes(codes) {
   let output = "";
   output += resolveAge(codes.charAt(0));
   output += resolveCategory(codes.charAt(1));
   output += resolveRegion(codes.charAt(2));
   output += resolveFrequency(codes.charAt(3));
   output += resolveSource(codes.charAt(4));
   
   if (output.length > 0) {
     return output.substring(0, output.length - 2);
   }
   return output;   
 }
 
 function resolveAge(code) {
   switch (code) {
     case 'A': return "age=archaic, ";
     case 'B': return "age=early, ";
     case 'C': return "age=classical, ";
     case 'D': return "age=late, ";
     case 'E': return "age=later, ";
     case 'F': return "age=medieval, ";
     case 'G': return "age=scholar, ";
     case 'H': return "age=modern, ";
     case 'X': return "";
     default: throw new Error("Unrecognized Age: " + code);
   }
 }
 
 function resolveCategory(code) {
   switch(code) {
     case 'A': return "cat=agr, ";
     case 'B': return "cat=bio, ";
     case 'D': return "cat=drm, ";
     case 'E': return "cat=ecc, ";
     case 'G': return "cat=grm, ";
     case 'L': return "cat=leg, ";
     case 'P': return "cat=poe, ";
     case 'S': return "cat=sci, ";
     case 'T': return "cat=tec, ";
     case 'W': return "cat=war, ";
     case 'Y': return "cat=myt, ";
     case 'X': return "";
     default: throw new Error("Unrecognized Category: " + code);
   }
 }
 
 function resolveRegion(code) {
   switch (code) {
     case 'A': return "reg=Aftrica, ";
     case 'B': return "reg=Britain, ";
     case 'C': return "reg=China, ";
     case 'D': return "reg=Scandinavia, ";
     case 'E': return "reg=Egypt, ";
     case 'F': return "reg=France/Gaul, ";
     case 'G': return "reg=Germany, ";
     case 'H': return "reg=Greece, ";
     case 'I': return "reg=Italy/Rome, ";
     case 'J': return "reg=India, ";
     case 'K': return "reg=Balkans, ";
     case 'N': return "reg=Netherlands, ";
     case 'P': return "reg=Persia, ";
     case 'Q': return "reg=NearEast, ";
     case 'R': return "reg=Russia, ";
     case 'S': return "reg=Spain/Iberia, ";
     case 'U': return "reg=Eastern Europe, ";
     case 'X': return "";
     default: throw new Error("Unrecognized Region: " + code);
   }
 }
 
 function resolveFrequency(code) {
   switch (code) {
     case 'A': return "frq=+++, ";
     case 'B': return "frq=++, ";
     case 'C': return "frq=+, ";
     case 'D': return "frq=-, ";
     case 'E': return "frq=--, ";
     case 'F': return "frq=---, ";
     case 'I': return "frq=inscription, ";
     case 'M': return "frq=grafffiti, ";
     case 'N': return "frq=Pliny, ";
     case 'X': return "";
     default: throw new Error("Unrecognized Frequency: " + code);
   }
 }
 
 function resolveSource(code) {
   switch (code) {
     case 'A':
     case 'B': return "src=Bee, ";
     case 'C': return "src=CAS, ";
     case 'D': return "src=Sex, ";
     case 'E': return "src=Ecc, ";
     case 'F': return "sec=Def, ";
     case 'G': return "src=G+L, ";
     case 'H': return "src=Ouvrard, ";
     case 'I': return "src=Leverett1845, ";
     case 'J':
     case 'K': return "src=Cal, ";
     case 'L': return "src=Levis1891, ";
     case 'M': return "src=Latham1980, ";
     case 'N': return "src=Nelson, ";
     case 'O': return "src=OLD, ";
     case 'P': return "src=Souter1949, ";
     case 'Q': return "src=other, ";
     case 'R': return "src=PlaterWhite"
     case 'S': return "src=L+S, ";
     case 'T': return "src=translation, ";
     case 'U': return "src=DuCange, ";
     case 'V': return "src=Saxo, ";
     case 'W': return "src=guess, ";
     case 'Z': return "src=user, ";
     case 'Y':
     case 'X': return "";
     default: throw new Error("Unrecognized Source: " + code);
   }
 }
 
 function cleanKind(_kind) {
   let kind = supertrim(_kind);
   const cut0 = kind.indexOf('(');
   const cut1 = kind.indexOf(')', cut0 + 1);
   if (cut0 !== -1 && cut1 > cut0 + 1) {
     const d = kind.charAt(cut0 + 1);
     if (d >= 0 && d <= 9) {
       kind = kind.substring(0,  cut0) + d + kind.substring(cut1 + 1);
     }
   }
   return kind;
 }
 
 process.on('exit', () => console.log("EXITED"))
 setInterval(() => {}, 1 << 30);
